### PR TITLE
Added rope to forestplot.py in forestplot()

### PR DESCRIPTION
**Pull request by GWeindel**
The new argument rope adds a shaded region of practical equivalence for all displayed variables in forestplot, shade is controlled by rope_alpha (default =.5). Contrary to posteriorplot rope can only be a tuple of length 2.
Not big but it is highly useful for me hence I assume it should be useful for others too.

### CODE REVIEW

**Comment by ahartikainen** [comment]
Sounds good. Can you post example?

**Comment by GWeindel** [comment]
@ahartikainen Sure, by the way it allowed me to spot an error in the ymax definition.
I post a reproducible example and link a plot I use.

Reproducible example : 
```
import pystan
import arviz as az

data = dict(N1=10, y1=2, N2=10, y2=6)

model = """//  Comparison of two groups with Binomial
data {
  int<lower=0> N1;
  int<lower=0> y1;
  int<lower=0> N2;
  int<lower=0> y2;
}
parameters {
  real<lower=0,upper=1> theta1;
  real<lower=0,upper=1> theta2;
}
model {
  theta1 ~ beta(1,1);
  theta2 ~ beta(1,1);
  y1 ~ binomial(N1,theta1);
  y2 ~ binomial(N2,theta2);
}
generated quantities {
  real oddsratio;
  oddsratio = (theta2/(1-theta2))/(theta1/(1-theta1));
}
"""

sm = pystan.StanModel(model_code=model)
fit = sm.sampling(data=data, iter=1000, chains=2)

az.plot_forest(fit, var_names=["theta1","theta2"], credible_interval=0.95, combined=True, rope=(-.1,.1),
               figsize=(7,7))

```
rope_example

**Comment by aloctavodia** [comment]
I like the idea, but it seems to me that the user should be allowed to set a rope per variable.

**Comment by GWeindel** [comment]
Sure I get it, I have no use for it but some might and consistency is defintely important. I will adapt the rope code of posteriorplot() to forestplot() when I have a bit time and perhaps generalize it to the ridgeplot option. Thanks for this highly useful package by the way.

**Comment by aloctavodia** [comment]
Great! And thank you for this contribution.

**Comment by GWeindel** [comment]
I severely rewrote my initial commit. Now rope has the same behavior as posteriorplot. I provide a reproductible example and a plot I could use.
example_az


Reproductible example : 
```
import pystan
import arviz as az
import stan_utility
import matplotlib.pyplot as plt

data = dict(N1=10, y1=2, N2=10, y2=6)

model = """//  Comparison of two groups with Binomial
data {
  int<lower=0> N1;
  int<lower=0> y1;
  int<lower=0> N2;
  int<lower=0> y2;
}
parameters {
  real<lower=0,upper=1> theta1;
  real<lower=0,upper=1> theta2;
}
model {
  theta1 ~ beta(1,1);
  theta2 ~ beta(1,1);
  y1 ~ binomial(N1,theta1);
  y2 ~ binomial(N2,theta2);
}
generated quantities {
  real oddsratio;
  oddsratio = (theta2/(1-theta2))/(theta1/(1-theta1));
}
"""

sm = pystan.StanModel(model_code=model)
fit = sm.sampling(data=data, iter=1000, chains=2)

rope = {'theta1' : [{'rope': (-.1, .1)}],
 'theta2' : [{'rope': (.2, .5)}],
 'oddsratio' : [{'rope': (10, 20)}]}

ropeList = (-.1, .1)

az.plot_forest(fit, var_names=["theta1","theta2"],credible_interval=0.95, combined=True, rope=rope, rope_values=True)
plt.show()
az.plot_forest(fit, var_names=["theta1","theta2"], credible_interval=0.95, combined=True, rope=ropeList,  figsize=(7,7))
plt.show()
```

**Comment by GWeindel** [comment]
However, I don't master classes so I might have introduced a lot of redundancy...

**Comment by GWeindel** [comment]
Faster reproducible and more complete example taken from the gallery : 

```
size = (1, 50)

data = {'normal': np.random.randn(*size),
    'gumbel': np.random.gumbel(size=size),
    'student t': np.random.standard_t(df=6, size=size),
    'exponential': np.random.exponential(size=size)}

rope = {'normal': [{"rope":(-.5,.5)}],
        'gumbel': [{"rope":(0,.5)}],
        'student t': [{"rope":(-.5,1)}],
        'exponential': [{"rope":(-.5,1)}]
       }

az.plot_forest(data, rope=(-1,1))
plt.show()
az.plot_forest(data, rope=rope)
plt.show()
az.plot_forest(data, rope=rope, rope_values=False)
plt.show()
```

**Comment by aloctavodia** [comment]
I like it. I think the rope should not be labeled here. I know it is for the posterior plot, by for that plot the hpd is also labeled.

**Comment by GWeindel** [comment]
You're right, I was hesitating whether to keep it.

File: arviz/plots/forestplot.py [hunk 1]
```diff
@@ -239,6 +243,20 @@ def labels_and_ticks(self):
             idxs.append(sub_idxs)
         return np.concatenate(labels), np.concatenate(idxs)
 
+    def display_multiple_ropes(self, rope, ax, y, linewidth, rope_var,markersize,xt_labelsize):
```
[the remark this answers is not in the document]
Comment (GWeindel) [reply arviz/plots/forestplot.py]: You're right, I remove them.

File: arviz/plots/forestplot.py [hunk 2]
```diff
@@ -302,8 +316,11 @@ def forestplot(
         else:
             qlist = [endpoint, 50, 100 - endpoint]
 
+        label, ticks = self.labels_and_ticks()
```
[the remark this answers is not in the document]
Comment (GWeindel) [reply arviz/plots/forestplot.py]: Right, and more efficient

File: arviz/plots/forestplot.py [hunk 3]
```diff
@@ -322,6 +339,17 @@ def forestplot(
         ax.set_title(
             "{:.1%} Credible Interval".format(credible_interval), fontsize=titlesize, wrap=True
         )
+        if rope is None or isinstance(rope, dict):
+            return
+        elif len(rope) == 2:
+            ax.axvspan(rope[0], rope[1], 0, self.y_max(), color='C2', alpha=.5)
+        else:
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (GWeindel) [reply arviz/plots/forestplot.py]: No credits taken, that's the exception message from posteriorplot :)

File: arviz/plots/forestplot.py [hunk 4]
```diff
@@ -239,6 +243,20 @@ def labels_and_ticks(self):
...
+            lw=linewidth * 2,
+            color="C2",
+            solid_capstyle="round",
+            zorder=0,
+            alpha=0.7,
+        )
+        return ax
+
+
     def ridgeplot(self, mult, xt_labelsize, linewidth, alpha, ax):
```
[the remark this answers is not in the document]
Comment (GWeindel) [reply arviz/plots/forestplot.py]: You're right it's unused. I add it to the PR

**Comment by canyon289** [comment]
@GWeindel Thank you for the quick edits. I'll do another round of review either today or tomorrow.

**Comment by canyon289** [comment]
@GWeindel Thanks for the fixes. Unfortunately our CI checks are broken at this commit so I'm running things manually as you make changes. Here's the errors I'm getting.

If you'd like you can rebase your changes on top of the current arviz master to get CI checks to work again. If you need help with this let me know.

Alternatively you can run 
```
./scripts/lint.sh
```
and it will generate the output below
> (arviz) tbayes@Users-MBP-2:~/repos/arviz$ ./scripts/lint.sh 
> ++ pwd
> + SRC_DIR=/Users/tbayes/repos/arviz
> + echo 'Checking documentation...'
> Checking documentation...
> + python -m pydocstyle --convention=numpy /Users/tbayes/repos/arviz/arviz/
> /Users/tbayes/repos/arviz/arviz/plots/forestplot.py:246 in public method `display_multiple_ropes`:
>         D210: No whitespaces allowed surrounding docstring text
> /Users/tbayes/repos/arviz/arviz/plots/forestplot.py:246 in public method `display_multiple_ropes`:
>         D400: First line should end with a period (not 'd')
> /Users/tbayes/repos/arviz/arviz/plots/forestplot.py:246 in public method `display_multiple_ropes`:
>         D401: First line should be in imperative mood ('Display', not 'Displays')

**Comment by ahartikainen** [comment]
LGTM. 

I fixed some annoying pylint errors.

**Comment by canyon289** [comment]
Does not LGTM yet,
Need test for new multi rope functionality. I can add one later today 

https://coveralls.io/builds/20848553/source?filename=arviz%2Fplots%2Fforestplot.py#L352

**Comment by canyon289** [comment]
Give me a couple more days, got caught up with other tasks

**Comment by canyon289** [comment]
Trying to add testing but messing up on arguments somewhere

```
import arviz as az
import matplotlib.pyplot as plt

az.style.use('arviz-darkgrid')

centered_data = az.load_arviz_data('centered_eight')

rope = {'mu': [{'rope': (-.1, .1)}], 'tau': [{'rope': (.2, .5)}]}
_, axes = az.plot_forest(centered_data,
                         var_names=['mu', 'tau'],
                         rope=rope)
axes[0].set_title('Estimated theta for eight schools model')
plt.show()
```

will continue working on this

**Comment by canyon289** [comment]
This implementation seems to break when each variable has multiple chains

**Combined = True** works
```
import arviz as az
import matplotlib.pyplot as plt
az.style.use('arviz-darkgrid')
centered_data = az.load_arviz_data('centered_eight')
rope = {'mu': [{'rope': (-.1, .1)}], 'tau': [{'rope': (.2, .5)}]}
_, axes = az.plot_forest(centered_data,
                         var_names=['mu', 'tau'],
                         combined=True,
                         rope=rope)
axes[0].set_title('Estimated theta for eight schools model')
plt.show()
```

**Combined = True** fails
```
import arviz as az
import matplotlib.pyplot as plt
az.style.use('arviz-darkgrid')
centered_data = az.load_arviz_data('centered_eight')
rope = {'mu': [{'rope': (-.1, .1)}], 'tau': [{'rope': (.2, .5)}]}
_, axes = az.plot_forest(centered_data,
                         var_names=['mu', 'tau'],
                         combined=False,
                         rope=rope)
axes[0].set_title('Estimated theta for eight schools model')
plt.show()
```

I have a couple suggested paths but happy to hear any others
1. Figure out how to make solution work with multiple chains
2. Provide a warning that per var rope won't work if multiple chains

I won't have time to code solution 1 if we choose to go that route, but can do solution 2.
Depending on which version we go for I can write the appropriate tests

**Comment by GWeindel** [comment]
I won't have the time (and perhaps sufficient skills) for solution 1. Hence I would suggest providing a warning.

**Comment by ahartikainen** [comment]
So what fails is

    ... label[ticks == y][0]

Which would mean that we don't have label for all `y`. Maybe try and if fail then skip that iteration?

**Comment by canyon289** [comment]
What happens is that ArviZ is trying to match the label with the line. When the chains are compressed label position matches the y position in the numpy array so one value is found, we get it out of the array in index zero.

In multi chain the ticks, and the y position, are off every so slightly, .175 if I remember correctly. The equality operator doesn't match anything and the array is empty, hence the index error.

I hope this not so great description yields some insight. I can provide more detail if needed

**Comment by ahartikainen** [comment]
Ok

We need y_group variable which is the mean of y locations per parameter (or value to be plotted). VarHandler probably needs to calculate these before hand, because all the yielding is disabling to do that later?

**Comment by ahartikainen** [comment]
Combined True

image


Combined False

image

**Comment by canyon289** [comment]
Now we just need to write a couple more tests and we're good!

**Comment by canyon289** [comment]
Thanks for the contribution @GWeindel

**Comment by GWeindel** [comment]
Great ! Thanks for the help with the multilple chains and the tests !

**Pull request merged**